### PR TITLE
[SPARK-30696][SQL] fromUTCtime and toUTCtime produced wrong result on Daylight Saving Time changes days

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
Before this patch, fromUTCtime and toUTCtime produced wrong result on Daylight Saving Time changes days
For example, in LA in 1960, timezone switch from UTC-7h to UTC-8h at 2AM in 1960-09-25 but previous version have the cutoff at 8AM

### Why are the changes needed?
correctness

### Does this PR introduce _any_ user-facing change?
yes, as correct the result produced by those two functions

### How was this patch tested?
Add test

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@dgd-contributor, please take a look at https://github.com/apache/spark/pull/32666/checks?check_run_id=2671403546 and enable Github actions for your forked repository.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 1]
```diff
@@ -787,4 +830,7 @@ class DateTimeUtilsSuite extends SparkFunSuite with Matchers with SQLHelper {
       }
     }
   }
+
+
+
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala:835]: Please remove unrelated changes.

Comment (dgd-contributor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: ok done!

**Comment by dgd-contributor** [comment]
Can anyone review this?

**Comment by dgd-contributor** [comment]
@cloud-fan could you review it?

**Comment by HyukjinKwon** [comment]
@MaxGekk should probably be one of the best persons to review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -839,8 +839,12 @@ object DateTimeUtils {
    * is valid for the `toZone` time zone, thus the local date-time may be adjusted.
    */
   def convertTz(micros: Long, fromZone: ZoneId, toZone: ZoneId): Long = {
-    val rebasedDateTime = getLocalDateTime(micros, toZone).atZone(fromZone)
-    instantToMicros(rebasedDateTime.toInstant)
+    val localDateTime = toJavaTimestamp(micros).toLocalDateTime
+    val zonedDateTimeFromZone = ZonedDateTime.of(localDateTime, fromZone)
+    val zonedDateTimeToZone = zonedDateTimeFromZone.withZoneSameInstant(toZone)
+    val localDateTimeConverted = zonedDateTimeToZone.toLocalDateTime
+    val timestamp = Timestamp.valueOf(localDateTimeConverted)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: BTW, seems like you're mixing JDK 7 and 8 libraries which will likely be correct.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: Mixing is possible via the rebase methods like `RebaseDateTime.rebaseGregorianToJulianMicros()` otherwise an implementation works incorrectly for ancient dates (before 1582-10-15) and timestamps (before 1900-01-01) because of different underlying calendar systems (Java 8 - Proleptic Gregorian calendar, Java 7 - the hybrid calendar Julian + Gregorian calendar), and JDK implementations - Java 8 uses more precise historical data.

Since you mix Java 8 methods and Java 7 `Timestamp.valueOf()`, highly likely your implementation will work incorrectly.

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: So what should I used to resolve this, any hints? Thanks

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: Sorry, can you explain more, I thought Timestamp.valueOf() is from java 1.8?
image

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: This method uses other methods of the `Timestamp` class that assumes the hybrid calendar. We described nuances of the hybrid and Proleptic Gregorian calendar (used since Spark 3.0) in the blog post https://databricks.com/blog/2020/07/22/a-comprehensive-look-at-dates-and-timestamps-in-apache-spark-3-0.html . I do believe that your changes can cause some issues in corner cases. I will try to allocate some time to find such cases slightly later.

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:846]: @MaxGekk any update?
